File: frontend/src/api/realtime/protocol.ts
/** 实时 SSE 事件流的 JSON 事件契约与解码，与 internal/realtime/protocol 保持一致。 */
import type {
  AgentRunBlockKind,
  AgentToolCallStatus,
  SyncHeads,
} from "../../../bindings/github.com/runforyou-ai/cervi/internal/appservice/models"

/** 运行过程流内容块的已知类型，取值与 AgentRunBlockKind 一致。 */
const blockKinds = new Set(["thinking", "content", "tool_call"])

/** 运行过程流工具调用的已知状态，取值与 AgentToolCallStatus 一致。 */
const toolCallStatuses = new Set(["queued", "running", "succeeded", "failed"])

/** 当前协议主版本，只有破坏性演进才提升。 */
export const realtimeProtocolVersion = 1

const int64Max = 9223372036854775807n

/** 运行过程流中的工具调用名称与状态，完整参数与结果经过程详情查询读取。 */
export type RunStreamToolCall = {
  name: string
  status: AgentToolCallStatus
  startedAt?: string
  completedAt?: string
}

/** 运行过程流中按位置排列的展示内容块。 */
export type RunStreamBlock = {
  id: string
  position: bigint
  kind: AgentRunBlockKind
  text: string
  toolCall?: RunStreamToolCall
}

/** 可按顺序应用到运行过程流快照的一条变更。 */
export type RunStreamOperation =
  | { kind: "upsert_block"; block: RunStreamBlock }
  | { kind: "append_block_text"; blockId: string; text: string }
  | { kind: "remove_blocks"; blockIds: string[] }
  | { kind: "append_candidate"; text: string }
  | { kind: "clear_candidate" }
  | { kind: "reset" }

/** 服务端经事件流下发的事件。 */
export type RealtimeServerFrame =
  | { type: "server_hello"; connectionId: string; syncHeads: SyncHeads }
  | { type: "visitor_hello"; connectionId: string }
  | { type: "ping" }
  | { type: "conversation_changed"; conversationId: string; version: bigint }
  | { type: "conversation_removed"; conversationId: string }
  | { type: "conversation_state_changed"; conversationId: string; version: bigint }
  | { type: "identity_profile_changed"; version: bigint }
  | {
      type: "run_stream_snapshot"
      runId: string
      streamId: string
      attempt: number
      sequence: bigint
      part: number
      partCount: number
      candidateContent: string
      blocks: RunStreamBlock[]
    }
  | {
      type: "run_stream_delta"
      runId: string
      streamId: string
      attempt: number
      baseSequence: bigint
      sequence: bigint
      operations: RunStreamOperation[]
    }
  | { type: "run_stream_ended"; runId: string }

/** 事件解码结果：未定义的事件种类忽略，主版本不一致与结构错误分别返回。 */
export type RealtimeServerFrameResult =
  | { status: "frame"; frame: RealtimeServerFrame }
  | { status: "ignored" }
  | { status: "unsupported_version" }
  | { status: "invalid"; reason: string }

type FrameData = Record<string, unknown>

/** 解码一条 SSE data 文本；先校验协议主版本，再校验事件种类、结构和字段类型。 */
export function decodeServerFrame(text: string): RealtimeServerFrameResult {
  try {
    const value: unknown = JSON.parse(text)
    if (!isFrameData(value) || (value.v !== undefined && typeof value.v !== "number")) {
      return { status: "invalid", reason: "frame envelope is malformed" }
    }
    if (value.v !== realtimeProtocolVersion) {
      return { status: "unsupported_version" }
    }
    const data = value.data ?? {}
    if (typeof value.type !== "string" || !isFrameData(data)) {
      return { status: "invalid", reason: "frame type or data is malformed" }
    }
    const frame = decodeServerData(value.type, data)
    return frame ? { status: "frame", frame } : { status: "ignored" }
  } catch (error) {
    return { status: "invalid", reason: error instanceof Error ? error.message : String(error) }
  }
}

/** 按事件种类读取事件数据，未定义的种类返回 undefined。 */
function decodeServerData(type: string, data: FrameData): RealtimeServerFrame | undefined {
  switch (type) {
    case "ping":
      return { type }
    case "server_hello": {
      // 探针校验和与身份资料版本是不透明比较值，保持字符串。
      const syncHeads = data.syncHeads
      if (!isFrameData(syncHeads) || !Number.isInteger(syncHeads.conversationCount)) {
        throw new Error("syncHeads is malformed")
      }
      return {
        type,
        connectionId: readString(data, "connectionId"),
        syncHeads: {
          conversationCount: syncHeads.conversationCount as number,
          conversationChecksum: readString(syncHeads, "conversationChecksum"),
          identityProfileVersion: readString(syncHeads, "identityProfileVersion"),
        },
      }
    }
    case "visitor_hello":
      return { type, connectionId: readString(data, "connectionId") }
    case "conversation_changed":
    case "conversation_state_changed":
      return { type, conversationId: readString(data, "conversationId"), version: readInt64(data, "version") }
    case "conversation_removed":
      return { type, conversationId: readString(data, "conversationId") }
    case "identity_profile_changed":
      return { type, version: readInt64(data, "version") }
    case "run_stream_snapshot":
      return {
        type,
        runId: readString(data, "runId"),
        streamId: readString(data, "streamId"),
        attempt: readInt(data, "attempt"),
        sequence: readInt64(data, "sequence"),
        part: readInt(data, "part"),
        partCount: readInt(data, "partCount"),
        candidateContent: typeof data.candidateContent === "string" ? data.candidateContent : "",
        blocks: readArray(data, "blocks").map(readBlock),
      }
    case "run_stream_delta":
      return {
        type,
        runId: readString(data, "runId"),
        streamId: readString(data, "streamId"),
        attempt: readInt(data, "attempt"),
        baseSequence: readInt64(data, "baseSequence"),
        sequence: readInt64(data, "sequence"),
        operations: readArray(data, "operations").map(readOperation),
      }
    case "run_stream_ended":
      return { type, runId: readString(data, "runId") }
    default:
      return undefined
  }
}

/** 读取一个运行过程流内容块，枚举值不在契约内时报错。 */
function readBlock(value: unknown): RunStreamBlock {
  if (!isFrameData(value)) {
    throw new Error("block is not an object")
  }
  const call = value.toolCall
  return {
    id: readString(value, "id"),
    position: readInt64(value, "position"),
    kind: readEnum(value, "kind", blockKinds) as AgentRunBlockKind,
    text: typeof value.text === "string" ? value.text : "",
    toolCall: isFrameData(call)
      ? {
          name: readString(call, "name"),
          status: readEnum(call, "status", toolCallStatuses) as AgentToolCallStatus,
          startedAt: typeof call.startedAt === "string" ? call.startedAt : undefined,
          completedAt: typeof call.completedAt === "string" ? call.completedAt : undefined,
        }
      : undefined,
  }
}

/** 读取一条增量操作，未定义的操作类型报错。 */
function readOperation(value: unknown): RunStreamOperation {
  if (!isFrameData(value)) {
    throw new Error("operation is not an object")
  }
  const kind = readString(value, "kind")
  switch (kind) {
    case "upsert_block":
      return { kind, block: readBlock(value.block) }
    case "append_block_text":
      return { kind, blockId: readString(value, "blockId"), text: readString(value, "text") }
    case "remove_blocks":
      return { kind, blockIds: readArray(value, "blockIds").map((id) => {
        if (typeof id !== "string") throw new Error("blockIds contains a non-string")
        return id
      }) }
    case "append_candidate":
      return { kind, text: readString(value, "text") }
    case "clear_candidate":
    case "reset":
      return { kind }
    default:
      throw new Error(`unsupported run stream operation ${kind}`)
  }
}

/** 判断值是否为 JSON 对象。 */
function isFrameData(value: unknown): value is FrameData {
  return typeof value === "object" && value !== null && !Array.isArray(value)
}

/** 读取字符串字段。 */
function readString(data: FrameData, key: string): string {
  const value = data[key]
  if (typeof value !== "string") {
    throw new Error(`${key} is not a string`)
  }
  return value
}

/** 读取数组字段，缺少字段时返回空数组。 */
function readArray(data: FrameData, key: string): unknown[] {
  const value = data[key]
  if (value === undefined || value === null) {
    return []
  }
  if (!Array.isArray(value)) {
    throw new Error(`${key} is not an array`)
  }
  return value
}

/** 读取非负整数字段。 */
function readInt(data: FrameData, key: string): number {
  const value = data[key]
  if (typeof value !== "number" || !Number.isInteger(value) || value < 0) {
    throw new Error(`${key} is not a non-negative integer`)
  }
  return value
}

/** 读取取值在已知集合内的字符串字段。 */
function readEnum(data: FrameData, key: string, values: Set<string>): string {
  const value = readString(data, key)
  if (!values.has(value)) {
    throw new Error(`${key} is not a known ${key} value`)
  }
  return value
}

/** 读取以十进制字符串传输的非负 64 位整数。 */
function readInt64(data: FrameData, key: string): bigint {
  const value = data[key]
  if (typeof value !== "string" || !/^(0|[1-9]\d*)$/.test(value) || BigInt(value) > int64Max) {
    throw new Error(`${key} is not an int64 string`)
  }
  return BigInt(value)
}

File: internal/realtime/protocol/protocol.go
// Package protocol 定义实时 SSE 事件流的 JSON 事件契约与编解码，TypeScript 端对应 frontend/src/api/realtime/protocol.ts。
package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/runforyou-ai/cervi/internal/appservice"
	"github.com/runforyou-ai/cervi/internal/domain"
)

// Version 是当前协议主版本，只有破坏性演进才提升。
const Version = 1

var (
	// ErrUnknownFrame 表示未定义的事件种类，接收方忽略该事件。
	ErrUnknownFrame = errors.New("realtime protocol: unknown event type")
	// ErrUnsupportedVersion 表示事件的协议主版本与当前版本不一致。
	ErrUnsupportedVersion = errors.New("realtime protocol: unsupported version")
)

// Type 定义事件种类。
type Type string

const (
	TypeServerHello              Type = "server_hello"
	TypeVisitorHello             Type = "visitor_hello"
	TypePing                     Type = "ping"
	TypeConversationChanged      Type = "conversation_changed"
	TypeConversationRemoved      Type = "conversation_removed"
	TypeConversationStateChanged Type = "conversation_state_changed"
	TypeConversationTyping       Type = "conversation_typing"
	TypeVisitorTyping            Type = "visitor_typing"
	TypeIdentityProfileChanged   Type = "identity_profile_changed"
	TypePinOrderChanged          Type = "pin_order_changed"
	TypeServiceAttention         Type = "service_attention"
	TypeRunStreamSnapshot        Type = "run_stream_snapshot"
	TypeRunStreamDelta           Type = "run_stream_delta"
	TypeRunStreamEnded           Type = "run_stream_ended"
)

// RunStreamOperationKind 定义运行过程流增量中的操作类型，与 agentruntime 的运行流操作一一对应。
type RunStreamOperationKind string

const (
	RunStreamUpsertBlock     RunStreamOperationKind = "upsert_block"
	RunStreamAppendBlockText RunStreamOperationKind = "append_block_text"
	RunStreamRemoveBlocks    RunStreamOperationKind = "remove_blocks"
	RunStreamAppendCandidate RunStreamOperationKind = "append_candidate"
	RunStreamClearCandidate  RunStreamOperationKind = "clear_candidate"
	RunStreamReset           RunStreamOperationKind = "reset"
)

// Frame 是可编码的实时事件。
type Frame interface {
	// FrameType 返回事件种类。
	FrameType() Type
}

// ServerHello 返回连接编号和订阅安装后读取的同步探针值；变更通知可能先于本事件到达。
type ServerHello struct {
	ConnectionID string               `json:"connectionId"`
	SyncHeads    appservice.SyncHeads `json:"syncHeads"`
}

// VisitorHello 返回网站访客事件流的连接编号；访客没有同步探针，重连后重新拉取目录与线程窗口。
type VisitorHello struct {
	ConnectionID string `json:"connectionId"`
}

// Ping 是服务端定期发送的心跳，客户端据此判断事件流仍然存活。
type Ping struct{}

// ConversationChanged 表示会话变到了指定版本。
type ConversationChanged struct {
	ConversationID string `json:"conversationId"`
	Version        int64  `json:"version,string"`
}

// ConversationRemoved 表示当前用户失去指定会话的阅读资格。
type ConversationRemoved struct {
	ConversationID string `json:"conversationId"`
}

// ConversationStateChanged 表示本人对会话的个人状态变到了指定版本。
type ConversationStateChanged struct {
	ConversationID string `json:"conversationId"`
	Version        int64  `json:"version,string"`
}

// ConversationTyping 表示会话中另一参与主体开始或停止输入，只驱动界面临时状态。
type ConversationTyping struct {
	ConversationID  string `json:"conversationId"`
	SenderSubjectID string `json:"senderSubjectId"`
	Active          bool   `json:"active"`
}

// VisitorTyping 表示客户会话中客服或 AI 员工正在准备回复，不携带内部身份编号。
type VisitorTyping struct {
	ConversationID string `json:"conversationId"`
	Active         bool   `json:"active"`
}

// IdentityProfileChanged 表示本人身份资料变到了指定版本。
type IdentityProfileChanged struct {
	Version int64 `json:"version,string"`
}

// PinOrderChanged 表示本人的个人置顶顺序变到了指定版本，置顶区需要整区重读。
type PinOrderChanged struct {
	Version int64 `json:"version,string"`
}

// ServiceAttention 提醒本人处理指定客户会话的客服处理周期，只驱动本地通知；负责人与队列状态以收件箱为准。
type ServiceAttention struct {
	ConversationID   string                        `json:"conversationId"`
	ServiceSessionID string                        `json:"serviceSessionId"`
	Reason           domain.ServiceAttentionReason `json:"reason"`
}

// RunStreamToolCall 是运行过程流中的工具调用名称、状态和起止时间，完整参数与结果经过程详情查询读取。
type RunStreamToolCall struct {
	Name        string                     `json:"name"`
	Status      domain.AgentToolCallStatus `json:"status"`
	StartedAt   *time.Time                 `json:"startedAt,omitempty"`
	CompletedAt *time.Time                 `json:"completedAt,omitempty"`
}

// RunStreamBlock 是运行过程流中按位置排列的展示内容块。
type RunStreamBlock struct {
	ID       string                   `json:"id"`
	Position int64                    `json:"position,string"`
	Kind     domain.AgentRunBlockKind `json:"kind"`
	Text     string                   `json:"text,omitempty"`
	ToolCall *RunStreamToolCall       `json:"toolCall,omitempty"`
}

// RunStreamOperation 是可按顺序应用到运行过程流快照的一条变更。
type RunStreamOperation struct {
	Kind     RunStreamOperationKind `json:"kind"`
	Block    *RunStreamBlock        `json:"block,omitempty"`
	BlockID  string                 `json:"blockId,omitempty"`
	BlockIDs []string               `json:"blockIds,omitempty"`
	Text     string                 `json:"text,omitempty"`
}

// RunStreamSnapshot 是运行过程流快照的一个分片；分片按 part 从 0 连续递增，收齐 partCount 个分片构成该序号上的完整快照。
type RunStreamSnapshot struct {
	RunID            string           `json:"runId"`
	StreamID         string           `json:"streamId"`
	Attempt          int              `json:"attempt"`
	Sequence         int64            `json:"sequence,string"`
	Part             int              `json:"part"`
	PartCount        int              `json:"partCount"`
	CandidateContent string           `json:"candidateContent,omitempty"`
	Blocks           []RunStreamBlock `json:"blocks"`
}

// RunStreamDelta 是运行过程流快照从起始序号到终止序号的增量，起始序号与当前快照序号不一致时接收方重新取快照。
type RunStreamDelta struct {
	RunID        string               `json:"runId"`
	StreamID     string               `json:"streamId"`
	Attempt      int                  `json:"attempt"`
	BaseSequence int64                `json:"baseSequence,string"`
	Sequence     int64                `json:"sequence,string"`
	Operations   []RunStreamOperation `json:"operations"`
}

// RunStreamEnded 表示本次执行尝试的运行过程流已结束，最终结果以持久查询为准。
type RunStreamEnded struct {
	RunID string `json:"runId"`
}

// FrameType 返回服务端 Hello 事件种类。
func (ServerHello) FrameType() Type { return TypeServerHello }

// FrameType 返回访客 Hello 事件种类。
func (VisitorHello) FrameType() Type { return TypeVisitorHello }

// FrameType 返回心跳事件种类。
func (Ping) FrameType() Type { return TypePing }

// FrameType 返回会话变更事件种类。
func (ConversationChanged) FrameType() Type { return TypeConversationChanged }

// FrameType 返回会话失权事件种类。
func (ConversationRemoved) FrameType() Type { return TypeConversationRemoved }

// FrameType 返回本人会话状态变更事件种类。
func (ConversationStateChanged) FrameType() Type { return TypeConversationStateChanged }

// FrameType 返回会话输入状态事件种类。
func (ConversationTyping) FrameType() Type { return TypeConversationTyping }

// FrameType 返回访客可见输入状态事件种类。
func (VisitorTyping) FrameType() Type { return TypeVisitorTyping }

// FrameType 返回身份资料变更事件种类。
func (IdentityProfileChanged) FrameType() Type { return TypeIdentityProfileChanged }

// FrameType 返回个人置顶顺序变更事件种类。
func (PinOrderChanged) FrameType() Type { return TypePinOrderChanged }

// FrameType 返回客服处理周期提醒事件种类。
func (ServiceAttention) FrameType() Type { return TypeServiceAttention }

// FrameType 返回运行过程流快照分片事件种类。
func (RunStreamSnapshot) FrameType() Type { return TypeRunStreamSnapshot }

// FrameType 返回运行过程流增量事件种类。
func (RunStreamDelta) FrameType() Type { return TypeRunStreamDelta }

// FrameType 返回运行过程流结束事件种类。
func (RunStreamEnded) FrameType() Type { return TypeRunStreamEnded }

// envelope 是事件在 SSE data 行中的外层结构。
type envelope struct {
	V    int             `json:"v"`
	Type Type            `json:"type"`
	Data json.RawMessage `json:"data,omitempty"`
}

// decoder 把事件数据解码为具体事件。
type decoder func(json.RawMessage) (Frame, error)

// decoders 是已定义的事件种类。
var decoders = map[Type]decoder{
	TypeServerHello:              decodeAs[ServerHello],
	TypeVisitorHello:             decodeAs[VisitorHello],
	TypePing:                     decodeAs[Ping],
	TypeConversationChanged:      decodeAs[ConversationChanged],
	TypeConversationRemoved:      decodeAs[ConversationRemoved],
	TypeConversationStateChanged: decodeAs[ConversationStateChanged],
	TypeConversationTyping:       decodeAs[ConversationTyping],
	TypeVisitorTyping:            decodeAs[VisitorTyping],
	TypeIdentityProfileChanged:   decodeAs[IdentityProfileChanged],
	TypePinOrderChanged:          decodeAs[PinOrderChanged],
	TypeServiceAttention:         decodeAs[ServiceAttention],
	TypeRunStreamSnapshot:        decodeAs[RunStreamSnapshot],
	TypeRunStreamDelta:           decodeAs[RunStreamDelta],
	TypeRunStreamEnded:           decodeAs[RunStreamEnded],
}

// Encode 把事件编码为带协议主版本的单行 JSON 文本。
func Encode(frame Frame) ([]byte, error) {
	data, err := json.Marshal(frame)
	if err != nil {
		return nil, fmt.Errorf("encode realtime event %s: %w", frame.FrameType(), err)
	}
	return json.Marshal(envelope{V: Version, Type: frame.FrameType(), Data: data})
}

// Decode 先校验协议主版本再按事件种类解码；只校验结构和类型，业务合法性由接收方判断。
func Decode(data []byte) (Frame, error) {
	var value envelope
	if err := json.Unmarshal(data, &value); err != nil {
		return nil, fmt.Errorf("decode realtime event: %w", err)
	}
	if value.V != Version {
		return nil, ErrUnsupportedVersion
	}
	decodeData, ok := decoders[value.Type]
	if !ok {
		return nil, ErrUnknownFrame
	}
	frame, err := decodeData(value.Data)
	if err != nil {
		return nil, fmt.Errorf("decode realtime event %s: %w", value.Type, err)
	}
	return frame, nil
}

// decodeAs 把事件数据解码为指定事件结构，缺少数据时返回零值事件。
func decodeAs[T Frame](data json.RawMessage) (Frame, error) {
	var frame T
	if len(data) == 0 {
		return frame, nil
	}
	err := json.Unmarshal(data, &frame)
	return frame, err
}
